### PR TITLE
go_router should allow setting requestFocus

### PR DESCRIPTION
**Pull request by hangyujin**
fixes: https://github.com/flutter/flutter/issues/129581

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/builder.dart [hunk 1]
```diff
@@ -63,6 +64,12 @@ class RouteBuilder {
   /// its history.
   final String? restorationScopeId;
 
+  /// Whether or not the navigator and it's new topmost route should request focus
+  /// when the new route is pushed onto the navigator.
+  /// 
+  /// Defaults to true.
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:70]: maybe add a link to navigator.requestfocus

File: packages/go_router/lib/src/builder.dart [hunk 2]
```diff
@@ -63,6 +64,12 @@ class RouteBuilder {
   /// its history.
   final String? restorationScopeId;
 
+  /// Whether or not the navigator and it's new topmost route should request focus
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:67]:
```suggestion
  /// Whether or not the navigator created by this builder and it's new topmost route should request focus
```

File: packages/go_router/lib/src/router.dart [hunk 3]
```diff
@@ -83,6 +83,7 @@ class GoRouter implements RouterConfig<RouteMatchList> {
     bool debugLogDiagnostics = false,
     GlobalKey<NavigatorState>? navigatorKey,
     String? restorationScopeId,
+    bool? requestFocus,
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:86]: can you add a doc comment in gorouter describe this parameter

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:86]: why is this nullable? I think we can default it to true?

File: packages/go_router/test/builder_test.dart [hunk 4]
```diff
@@ -424,6 +424,7 @@ class _BuilderTestWidget extends StatelessWidget {
       restorationScopeId: null,
       observers: <NavigatorObserver>[],
       onPopPageWithRouteMatch: (_, __, ___) => false,
+      requestFocus: null,
```
Comment (chunhtai) [inline packages/go_router/test/builder_test.dart:427]: why this change?

File: packages/go_router/lib/src/router.dart (hunk 3 above)
Comment (hangyujin) [reply packages/go_router/lib/src/router.dart:86]: done

File: packages/go_router/lib/src/builder.dart (hunk 1 above)
Comment (hangyujin) [reply packages/go_router/lib/src/builder.dart:70]: done

File: packages/go_router/lib/src/builder.dart [hunk 5]
```diff
@@ -63,6 +64,14 @@ class RouteBuilder {
   /// its history.
   final String? restorationScopeId;
 
+  /// Whether or not the navigator created by this builder and it's new topmost route should request focus
+  /// when the new route is pushed onto the navigator.
+  ///
+  /// Defaults to true.
+  ///
+  /// {@macro flutter.widgets.navigator.requestFocus}
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:72]: Does macro work in package?

Comment (hangyujin) [reply packages/go_router/lib/src/builder.dart:72]: deleted.
 looks like it doesn't work
